File: android/src/main/java/io/refiner/rn/RNRefinerModule.java

package io.refiner.rn;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.modules.core.RCTNativeAppEventEmitter;

import java.util.HashMap;
import java.util.LinkedHashMap;

import io.refiner.Refiner;
import io.refiner.rn.utils.MapUtil;
import kotlinx.serialization.json.Json;
import kotlinx.serialization.json.JsonObject;

public class RNRefinerModule extends ReactContextBaseJavaModule {

    private final ReactApplicationContext reactContext;

    public RNRefinerModule(ReactApplicationContext reactContext) {
        super(reactContext);
        this.reactContext = reactContext;
    }

    @Override
    public String getName() {
        return "RNRefiner";
    }

    @ReactMethod
    public void initialize(String projectId, Boolean debugMode) {
        if (projectId != null) {
            Refiner.INSTANCE.initialize(reactContext, projectId, debugMode);
            registerCallbacks();
        }
    }

    @ReactMethod
    public void setProject(String projectId) {
        if (projectId != null) {
            Refiner.INSTANCE.setProject(projectId);
        }
    }

    @ReactMethod
    public void identifyUser(String userId, ReadableMap userTraits, String locale, String signature) {
        LinkedHashMap<String, Object> userTraitsMap = null;
        if (userTraits != null) {
            userTraitsMap = new LinkedHashMap<>(MapUtil.toMap(userTraits));
        }
        if (userId != null) {
            Refiner.INSTANCE.identifyUser(userId, userTraitsMap, locale, signature);
        }
    }

    @ReactMethod
    public void resetUser() {
        Refiner.INSTANCE.resetUser();
    }

    @ReactMethod
    public void trackEvent(String eventName) {
        if (eventName != null) {
            Refiner.INSTANCE.trackEvent(eventName);
        }
    }

    @ReactMethod
    public void trackScreen(String screenName) {
        if (screenName != null) {
            Refiner.INSTANCE.trackScreen(screenName);
        }
    }

    @ReactMethod
    public void ping() {
        Refiner.INSTANCE.ping();
    }

    @ReactMethod
    public void showForm(String formUuid, boolean force) {
        if (formUuid != null) {
            Refiner.INSTANCE.showForm(formUuid, force);
        }
    }

    @ReactMethod
    public void dismissForm(String formUuid) {
        if (formUuid != null) {
            Refiner.INSTANCE.dismissForm(formUuid);
        }
    }

    @ReactMethod
    public void closeForm(String formUuid) {
        if (formUuid != null) {
            Refiner.INSTANCE.closeForm(formUuid);
        }
    }

    @ReactMethod
    public void addToResponse(ReadableMap contextualData) {
        HashMap<String, Object> contextualDataMap = null;
        if (contextualData != null) {
            contextualDataMap = new HashMap<>(MapUtil.toMap(contextualData));
        }
        Refiner.INSTANCE.addToResponse(contextualDataMap);
    }

    @ReactMethod
    public void startSession() {
        Refiner.INSTANCE.startSession();
    }

    @Deprecated
    @ReactMethod
    public void attachToResponse(ReadableMap contextualData) {
        addToResponse(contextualData);
    }

    @ReactMethod
    public void addListener(String eventName) {
        // Keep: Required for RN built in Event Emitter Calls.
    }

    @ReactMethod
    public void removeListeners(Integer count) {
        // Keep: Required for RN built in Event Emitter Calls.
    }

    private void registerCallbacks() {
        Refiner.INSTANCE.onBeforeShow((formId, formConfig) -> {
            String config = Json.Default.encodeToString(JsonObject.Companion.serializer(), (JsonObject) formConfig);
            WritableMap params = Arguments.createMap();
            params.putString("formId", formId);
            params.putString("formConfig", config);
            sendEvent("onBeforeShow", params);
            return null;
        });

        Refiner.INSTANCE.onShow((formId) -> {
            WritableMap params = Arguments.createMap();
            params.putString("formId", formId.toString());
            sendEvent("onShow", params);
            return null;
        });

        Refiner.INSTANCE.onClose((formId) -> {
            WritableMap params = Arguments.createMap();
            params.putString("formId", formId.toString());
            sendEvent("onClose", params);
            return null;
        });

        Refiner.INSTANCE.onDismiss((formId) -> {
            WritableMap params = Arguments.createMap();
            params.putString("formId", formId.toString());
            sendEvent("onDismiss", params);
            return null;
        });

        Refiner.INSTANCE.onComplete((formId, formData) -> {
            String data = Json.Default.encodeToString(JsonObject.Companion.serializer(), (JsonObject) formData);
            WritableMap params = Arguments.createMap();
            params.putString("formId", formId.toString());
            params.putString("formData", data);
            sendEvent("onComplete", params);
            return null;
        });

        Refiner.INSTANCE.onNavigation((formId, formElement, progress) -> {
            String element = Json.Default.encodeToString(JsonObject.Companion.serializer(), (JsonObject) formElement);
            String pro = Json.Default.encodeToString(JsonObject.Companion.serializer(), (JsonObject) progress);

            WritableMap params = Arguments.createMap();
            params.putString("formId", formId);
            params.putString("formElement", element);
            params.putString("progress", pro);
            sendEvent("onNavigation", params);
            return null;
        });
    }

    private void sendEvent(String eventName, Object params) {
        try {
            getReactApplicationContext().getJSModule(RCTNativeAppEventEmitter.class).emit(eventName, params);
//            Log.d(TAG, "Sending event " + eventName);
        } catch (Throwable t) {
//            Log.e(TAG, t.getLocalizedMessage());
        }
    }
}

File: android/src/main/java/io/refiner/rn/RNRefinerModule.kt
package io.refiner.rn

import com.facebook.react.bridge.*
import com.facebook.react.modules.core.DeviceEventManagerModule
import io.refiner.Refiner
import io.refiner.rn.utils.MapUtil
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject

class RNRefinerModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {

    companion object {
        const val NAME = "RNRefiner"
        private val isNewArchitecture = BuildConfig.IS_NEW_ARCHITECTURE_ENABLED
    }

    init {
        android.util.Log.d("RNRefinerModule", "RNRefinerModule instantiated, isNewArch: $isNewArchitecture")
    }

    override fun getName(): String = NAME

    @ReactMethod
    @Suppress("UNUSED_PARAMETER")
    fun setArchitectureInfo(isNewArch: Boolean) {
        // Architecture is determined at build time via BuildConfig.IS_NEW_ARCHITECTURE_ENABLED
        // This method is kept for compatibility but doesn't change the architecture
    }

    @ReactMethod
    fun getArchitectureInfo(promise: Promise) {
        promise.resolve(isNewArchitecture)
    }

    @ReactMethod
    fun detectArchitecture(promise: Promise) {
        promise.resolve(isNewArchitecture)
    }

    @ReactMethod
    fun initialize(projectId: String?, debugMode: Boolean?) {
        android.util.Log.d("RNRefinerModule", "initialize called with projectId: $projectId, debugMode: $debugMode")
        projectId?.let { id ->
            Refiner.initialize(reactApplicationContext.applicationContext, id, debugMode)
            registerCallbacks()
        }
    }

    @ReactMethod
    fun setProject(projectId: String?) {
        projectId?.let { id ->
            Refiner.setProject(id)
        }
    }

    @ReactMethod
    fun identifyUser(
        userId: String?,
        userTraits: ReadableMap?,
        locale: String?,
        signature: String?,
        writeOperation: String?
    ) {
        val userTraitsMap = userTraits?.let { MapUtil.toMap(it) }
        
        userId?.let { id ->
            val operation = writeOperation?.let { Refiner.WriteOperation.valueOf(it) } 
                ?: Refiner.WriteOperation.APPEND
            Refiner.identifyUser(id, userTraitsMap, locale, signature, operation)
        }
    }

    @ReactMethod
    fun setUser(
        userId: String?,
        userTraits: ReadableMap?,
        locale: String?,
        signature: String?
    ) {
        val userTraitsMap = userTraits?.let { MapUtil.toMap(it) }
        
        userId?.let { id ->
            Refiner.setUser(id, userTraitsMap, locale, signature)
        }
    }

    @ReactMethod
    fun resetUser() {
        Refiner.resetUser()
    }

    @ReactMethod
    fun trackEvent(eventName: String?) {
        eventName?.let { name ->
            Refiner.trackEvent(name)
        }
    }

    @ReactMethod
    fun trackScreen(screenName: String?) {
        screenName?.let { name ->
            Refiner.trackScreen(name)
        }
    }

    @ReactMethod
    fun ping() {
        Refiner.ping()
    }

    @ReactMethod
    fun showForm(formUuid: String?, force: Boolean) {
        android.util.Log.d("RNRefinerModule", "showForm called with formUuid: $formUuid, force: $force")
        formUuid?.let { uuid ->
            Refiner.showForm(uuid, force)
        }
    }

    @ReactMethod
    fun dismissForm(formUuid: String?) {
        formUuid?.let { uuid ->
            Refiner.dismissForm(uuid)
        }
    }

    @ReactMethod
    fun closeForm(formUuid: String?) {
        formUuid?.let { uuid ->
            Refiner.closeForm(uuid)
        }
    }

    @ReactMethod
    fun addToResponse(contextualData: ReadableMap?) {
        val contextualDataMap = contextualData?.let { MapUtil.toMap(it) }
        Refiner.addToResponse(contextualDataMap)
    }

    @ReactMethod
    fun startSession() {
        Refiner.startSession()
    }

    @Deprecated("Use addToResponse instead", ReplaceWith("addToResponse(contextualData)"))
    @ReactMethod
    fun attachToResponse(contextualData: ReadableMap?) {
        addToResponse(contextualData)
    }

    @ReactMethod
    @Suppress("UNUSED_PARAMETER")
    fun addListener(eventName: String?) {
        // Keep: Required for RN built in Event Emitter Calls.
    }

    @ReactMethod
    @Suppress("UNUSED_PARAMETER")
    fun removeListeners(count: Int?) {
        // Keep: Required for RN built in Event Emitter Calls.
    }

    private fun registerCallbacks() {
        Refiner.onBeforeShow { formId, formConfig ->
            val config = Json.Default.encodeToString(JsonObject.serializer(), formConfig as JsonObject)
            val params = Arguments.createMap().apply {
                putString("formId", formId)
                putString("formConfig", config)
            }
            sendEvent("onBeforeShow", params)
        }

        Refiner.onShow { formId ->
            val params = Arguments.createMap().apply {
                putString("formId", formId.toString())
            }
            sendEvent("onShow", params)
        }

        Refiner.onClose { formId ->
            val params = Arguments.createMap().apply {
                putString("formId", formId.toString())
            }
            sendEvent("onClose", params)
        }

        Refiner.onDismiss { formId ->
            val params = Arguments.createMap().apply {
                putString("formId", formId.toString())
            }
            sendEvent("onDismiss", params)
        }

        Refiner.onComplete { formId, formData ->
            val data = Json.Default.encodeToString(JsonObject.serializer(), formData as JsonObject)
            val params = Arguments.createMap().apply {
                putString("formId", formId.toString())
                putString("formData", data)
            }
            sendEvent("onComplete", params)
        }

        Refiner.onNavigation { formId, formElement, progress ->
            val element = Json.Default.encodeToString(JsonObject.serializer(), formElement as JsonObject)
            val pro = Json.Default.encodeToString(JsonObject.serializer(), progress as JsonObject)

            val params = Arguments.createMap().apply {
                putString("formId", formId)
                putString("formElement", element)
                putString("progress", pro)
            }
            sendEvent("onNavigation", params)
        }

        Refiner.onError { message ->
            val params = Arguments.createMap().apply {
                putString("message", message.toString())
            }
            sendEvent("onError", params)
        }
    }

    private fun sendEvent(eventName: String, params: WritableMap?) {
        try {
            reactApplicationContext
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
                .emit(eventName, params)
        } catch (t: Throwable) {
            // Log error if needed
        }
    }
} 
